File: JavaClient/src/main/java/com/security/crypto/Handshake/DHkeyExchange.java
package com.security.crypto.Handshake;

import com.security.crypto.Ciphers.AES.AesECB;
import com.security.crypto.Ciphers.AES.HMacAlgoProvider;
import com.security.crypto.Ciphers.RSA.RSA_PKCS1;
import com.security.crypto.Configuration.*;
import com.security.crypto.IOSocket.IOSynAck;
import com.security.crypto.IOSocket.IOTransport;
import com.security.crypto.KeyManager.KeyHandler;

import java.io.IOException;
import java.math.BigInteger;
import java.util.StringJoiner;


@SuppressWarnings("ALL")
public final class DHkeyExchange extends IOSynAck {

    private IOTransport SocketChanel;
    private RandomGenerator Genarator;
    private final KeyHandler keystore;
    private AesECB aesNoIVParams;
    private RSA_PKCS1 rsa_pkcs1;
    private CookieGen cookie;
    private StringJoiner Ciphers;
    private StringJoiner Diggest;
    private StringJoiner CurrentDiggest;

    public DHkeyExchange(IOTransport SocketChanel, KeyHandler keystore) {
        this.SocketChanel = SocketChanel;
        this.cookie = new CookieGen();
        this.Genarator = new RandomGenerator();
        this.keystore = keystore;
        this.aesNoIVParams = new AesECB();
        this.rsa_pkcs1 = new RSA_PKCS1();
        Ciphers = new StringJoiner(",");
        Diggest = new StringJoiner(",");
        CurrentDiggest = new StringJoiner(",");
    }

    @Override
    public void SendPlainMessage() throws IOException {
        //send PublicKey to Server
        JSonObject ObjToSend = new JSonObject();
        ObjToSend.PlainMessage = Properties.SYN;
        ObjToSend.PseudoNumber = Genarator.pseudorandom();
        String toSend = JSonParse.WriteObject(ObjToSend);
        SocketChanel.SendMessage(toSend);
    }

    public void ReceiveServerCertificate() throws Exception {
        JSonObject receivedObj = JSonParse.ReadObject(SocketChanel.receiveMessage());
        String timestamp = Genarator.pseudorandom();
        if (!receivedObj.PlainMessage.equals(Properties.SYN_ACK) ||
                !receivedObj.PseudoNumber.equals(timestamp))
            throw new Exception("Server Cannot Be Verified");
        else
            this.keystore.SaveCertificate(receivedObj.CertPemFormat);
        cookie.setCookieServer(receivedObj.CookieServer);
        return;
    }

    public void ResendCookieServer() throws IOException {
        JSonObject ObjToSend = new JSonObject();
        ObjToSend.PlainMessage = Properties.Replay;
        ObjToSend.PseudoNumber = Genarator.pseudorandom();
        ObjToSend.CookieServer = this.cookie.getCookieServer();
        String toSend = JSonParse.WriteObject(ObjToSend);
        SocketChanel.SendMessage(toSend);
        return;
    }


    public void SendPublicValue() throws IOException {
        BigInteger ServerPublicPrimeNumber = Genarator.GeneratePublicPrimeNumber();
        try {
            JSonObject ObjToSend = new JSonObject();
            ObjToSend.PseudoNumber = Genarator.pseudorandom();
            ObjToSend.ClientEncryptedPrimeNumber = this.rsa_pkcs1.RsaEncrypt(
                    keystore.loadRemoteServerPublicKey(), ServerPublicPrimeNumber.toString());

            String toSend = JSonParse.WriteObject(ObjToSend);
            SocketChanel.SendMessage(toSend);
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    public void ReceivePublicValue() throws Exception {
        JSonObject receivedObj = JSonParse.ReadObject(SocketChanel.receiveMessage());
        if (!receivedObj.PseudoNumber.equals(Genarator.pseudorandom()))
            throw new Exception("Server Cannot Be Verified Possible Replay Attack");

        BigInteger sessionResult = Genarator.SessionGenerator(receivedObj.ServerPrimeNumber);
        keystore.ProduceCipherKey(sessionResult.toString());//Produce and save Ciphers Key from The given Session Result
        keystore.ProduceIntegrityKey(sessionResult.toString());//Produce and save Integrity Key from The given Session Result
        System.out.println(sessionResult.toString());
        return;
    }

    public void SendCipherSuites() throws Exception {
        JSonObject ObjToSend = new JSonObject();
        StringJoiner joiner = new StringJoiner("|");
        Ciphers.add(Properties.AES_ECB).add(Properties.AES_CBC);
        Diggest.add(Properties.MD5).add(Properties.sha1).add(Properties.MACSHA_256);
        CurrentDiggest.add(Properties.MACSHA_256);
        joiner.add(Ciphers.toString()).add(Diggest.toString()).add(CurrentDiggest.toString());

        // System.out.println(joiner.toString());
        ObjToSend.PseudoNumber = Genarator.pseudorandom();
        ObjToSend.CipherSuites = joiner.toString();
        ObjToSend.HmacHash = HMacAlgoProvider.HmacSign(joiner.toString(), keystore.loadRemoteIntegrityKey(), CurrentDiggest.toString());

        String toSend = JSonParse.WriteObject(ObjToSend);
        SocketChanel.SendMessage(toSend);
    }

    public CiphersForUse ReceiveCipherSuites() throws Exception {
        JSonObject receivedObj = JSonParse.ReadObject(SocketChanel.receiveMessage());
        if (!receivedObj.PseudoNumber.equals(Genarator.pseudorandom()) ||
                !HMacAlgoProvider.HmacVerify(receivedObj.CipherSuites, keystore.loadRemoteIntegrityKey(), receivedObj.HmacHash, CurrentDiggest.toString()))
            throw new Exception("Server Cannot Be Verified Possible Replay Attack");


        String SelectedCiphers = receivedObj.CipherSuites;
        System.out.println(SelectedCiphers);
        String[] parts = null;

        if (SelectedCiphers.contains("|"))
            parts = SelectedCiphers.split("\\|");
        else
            throw new IllegalArgumentException("String " + SelectedCiphers + " does not contain |");

        //   System.out.println(parts.toString());
        String CipherAlgo = parts[0];
        String HashAlgo = parts[1];
        return new CiphersForUse(CipherAlgo, HashAlgo);


    }

}

File: SecureUWPClient/SecureUWPClient/Handshake/DHkeyExchange.cs
using System;
using System.Threading.Tasks;
using SecureUWPChannel.Interfaces;
using SecureUWPChannel.Serialization;
using SecureUWPChannel.Prooperties;
using SecureUWPClient.Configuration;
using SecureUWPClient.KeyManager;
using SecureUWPClient.Ciphers.RSA;
using System.Diagnostics;
using SecureUWPClient.Ciphers.AES;

namespace SecureUWPChannel.IOTransport
{
    public class DHkeyExchange :IAsyncSynAck
    {
        private IOMulticastAndBroadcast ActivitySocket;
        private PrimeNumberGenerator Genarator;
        private KeyManagerImp keystore;
        private Cookie cOokie;
        private RSA_PKCS1 rsa_pkcs1;
        private JSonObject ReadObj;
        private JSonObject WriteObj;


        private string Ciphers;
        private string Diggest;
        private string CurrentDiggest;

        public DHkeyExchange(IOMulticastAndBroadcast ActivitySocket, KeyManagerImp keystore)
        {
            this.keystore = keystore;
            this.Genarator = new PrimeNumberGenerator();
            this.cOokie = new Cookie();
            this.rsa_pkcs1 = new RSA_PKCS1();
            this.ActivitySocket = ActivitySocket;
            Intialize();
        }
       

        public void Intialize()
        {
            ReadObj = new JSonObject();
            WriteObj = new JSonObject();
        }

        public override async Task SendPlainMessage()
        {
            WriteObj.PlainMessage = SampleConfiguration.SYN;
            WriteObj.PseudoNumber = Genarator.pseudorandom();

            String toSend = JsonParse.WriteObject(WriteObj);
            await this.ActivitySocket.send(toSend);

        }

        public override async Task ReceiveServerCertificate()
        {
            ReadObj = JsonParse.ReadObject(await this.ActivitySocket.read()); 
            String timestamp = Genarator.pseudorandom();
            if (!ReadObj.PlainMessage.Equals(SampleConfiguration.SYN_ACK) ||
                    !timestamp.Contains(ReadObj.PseudoNumber))
                throw new Exception("Server Cannot Be Verified");

            /*  var t=Task.Run(async () => {
                 await keystore.SaveCertificate(ReadObj.CertPemFormat);
             });
            t.Wait();
            t=Task.Run(async () => {
                 keystore.SaveServerPublicKey(await keystore.LoadCertificate());
             });
            t.Wait();*/
            await keystore.SaveCertificate(ReadObj.CertPemFormat);
            keystore.SaveServerPublicKey(await keystore.LoadCertificate());
            cOokie.CookieServer=ReadObj.CookieServer;
        }

        public override async Task ResendCookieServer()
        {
            WriteObj = new JSonObject();
            WriteObj.PlainMessage = SampleConfiguration.Replay;
            WriteObj.PseudoNumber = Genarator.pseudorandom();
            WriteObj.CookieServer = cOokie.CookieServer;

            String toSend = JsonParse.WriteObject(WriteObj);
            await this.ActivitySocket.send(toSend);
            return;
        }


        public override async Task SendPublicValue()
        {
            String ServerPublicPrimeNumber = Genarator.GetClientPublicNumber();
            try
            {
                WriteObj = new JSonObject();
                WriteObj.PseudoNumber = Genarator.pseudorandom();
                String encrypted= await rsa_pkcs1.RsaEncrypt(await keystore.LoadPublicKey(),ServerPublicPrimeNumber);
                WriteObj.ClientEncryptedPrimeNumber = encrypted;

                String toSend = JsonParse.WriteObject(WriteObj);
                await this.ActivitySocket.send(toSend);
                return;
            }
            catch (Exception e)
            {
                Debug.WriteLine(e.Message);
            }
        }


        public override async Task ReceivePublicValue()
        {
            ReadObj = JsonParse.ReadObject(await this.ActivitySocket.read());
            String timestamp = Genarator.pseudorandom();
            if (!timestamp.Contains(ReadObj.PseudoNumber))
                throw new Exception("Server Cannot Be Verified Possible Replay Attack");

            String sessionResult = Genarator.SessionDHGenerator(ReadObj.ServerPrimeNumber);
            keystore.ProduceCipherKey(sessionResult);//Produce and save Ciphers Key from The given Session Result
            keystore.ProduceIntegrityKey(sessionResult);//Produce and save Integrity Key from The given Session Result
            Debug.WriteLine(sessionResult);
            return;
        }

        public override async Task SendCipherSuites()
        {
            WriteObj = new JSonObject();

             Ciphers = StringUtils.Joiner(",",SampleConfiguration.AES_ECB, SampleConfiguration.AES_CBC);
             Diggest = StringUtils.Joiner(",", SampleConfiguration.MD5, SampleConfiguration.sha1, SampleConfiguration.SHA_256);
             CurrentDiggest = StringUtils.Joiner(",", SampleConfiguration.MACSHA_256);
            string joiner = StringUtils.Joiner("|", Ciphers, Diggest, CurrentDiggest);

            WriteObj.PseudoNumber = Genarator.pseudorandom();
            WriteObj.CipherSuites = joiner;
            WriteObj.HmacHash = HMacAlgoProvider.CreateHMAC(joiner, await keystore.LoadIntegrityKey(), CurrentDiggest);

            String toSend = JsonParse.WriteObject(WriteObj);
            await this.ActivitySocket.send(toSend);
            return;
        }

        public override async Task<CiphersForUse> ReceiveCipherSuites()
        {
            ReadObj = JsonParse.ReadObject(await this.ActivitySocket.read());
            String timestamp = Genarator.pseudorandom(); 
            if (!timestamp.Contains(ReadObj.PseudoNumber) ||
                !HMacAlgoProvider.VerifyHMAC(ReadObj.CipherSuites, await keystore.LoadIntegrityKey(), ReadObj.HmacHash, CurrentDiggest))
                throw new Exception("Server Cannot Be Verified Possible Replay Attack");


            String SelectedCiphers = ReadObj.CipherSuites;
            Debug.WriteLine(SelectedCiphers);
            String[] parts = null;

            if (SelectedCiphers.Contains("|"))
                parts = SelectedCiphers.Split('|');
            else
                throw new Exception("String " + SelectedCiphers + " does not contain |");

            //   System.out.println(parts.toString());
            String CipherAlgo = StringUtils.Replace("/","_",parts[0]);
            String HashAlgo = parts[1];
            return new CiphersForUse(CipherAlgo, HashAlgo);
        }

    }
}
